### DISCUSSION TITLE
v4:  pagination for Issues + PullRequests for a specific repository

### DISCUSSION THREAD

**Discussion by kttkjl** [community/community#24398] [Apps, API and Webhooks]
<p>Github API V3 states</p>
<p>“a pull request is an issue, but an issue is not always a pull request, please check for the pull_request key”</p>
<p>However, it seems for V4, Issues and PullRequests are a different category, I can paginate each individually, but I can’t wrap my head around getting pagination to work with both of them together … is there a way to do it like V3?</p>
<p>Say I grab 50 issues, Issue AND PR, and then paginate with the cursor </p>
<p>… or should I just stick with v3 for this?</p>

**Comment by lee-dohm** [comment]
<p>Do you have a sample query to start from? I’m not sure where the problem is from your description.</p>
<p>Thanks.</p>

**Comment by kttkjl** [comment]
<p>Whoops, really sorry about the late reply, I thought I’d get a notification on github </p>
<p>So say I wanted to query issues by looking at the repo owner, and the name - this is V4 of the API</p>
<pre><code>query get20Issues($owner: String!, $name: String!, $cursor: String) {
    repository(owner: $owner, name: $name) {
      issues(
        first: 20
        after: $cursor
        orderBy: { field: CREATED_AT, direction: DESC }
      ) {
        nodes {
          number
          body: bodyText
          title
          state
          labels(first: 5) {
            nodes {
              name
              color
            }
          }
        }
        pageInfo {
          endCursor
        }
      }
    }
  }
</code></pre>
<p>What it returns is an array of issue nodes … all fine and good, but what it lacks is that <strong>issues are separate from pull-requests </strong>… I’m guessing that’s the design decision, but it’s different from that of V3, where each issue object returned had a <strong>pull_request</strong> key, signifying that this issue is also a pull request.</p>
<p>However, that means for V4 we would have to query for both issues and pull-requests keep the <strong>pageInfo </strong> for both … </p>
<pre><code>query getNIssuePr($owner: String!, $name: String!, $issue_cursor: String, $pr_cursor: String, $num:Int) {
    repository(owner: $owner, name: $name) {
      issues(
        first: $num
        after: $issue_cursor
        orderBy: { field: CREATED_AT, direction: DESC }
      ) {
        nodes {
          number
          ...
          labels(first: 5) {
            nodes {
              name
              color
            }
          }
        }
        pageInfo {
          endCursor
        }
      }
      pullRequests(
        first: $num
        after: $pr_cursor
        orderBy: { field: CREATED_AT, direction: DESC }
      ) {
        nodes {
          number
...
          labels(first: 5) {
            nodes {
              name
              color
            }
          }
        }
        pageInfo {
          endCursor
        }
      }
    }
  }
</code></pre>
<p>Which will give us 20 of the most recent of both issues and PRs</p>
<p>Now if we were to call V3, then we’ll go ahead and grab the most recent 40 issues - SOME of them PR, but not 50% always, but they’ll always be the most recent 40 issues/PR, and we can tell via the  <strong>pull_request</strong> key.</p>
<p>Going back to V4, the particular repo might have the most recent 30 issues be normal issues, and 31-50 being PRs, by my 2nd snippet, we’ll be grabbing 0-19 as issue, then 31-50 as PR, skipping 20-29</p>
<p>So I was wondering if I could do the same thing as api V3 in api V4</p>
<p>Thanks!</p>

**Comment by lee-dohm** [comment]
<p>There isn’t a connection that offers pagination, but there is the <code>issueOrPullRequest</code> field that allows you to query by number. So you could get the latest issue and the latest pull request, pick the higher number and then count backwards from there to get the latest 40 issues or pull requests.</p>
<p>I hope that helps.</p>

**Comment by kttkjl** [comment]
<p>Please pardon my inexperience, so that would mean, I’d have to to multiple queries then?</p>
<ol>
<li>
<p>Grab latest number from issue or PR</p>
</li>
<li>
<p>a forEach loop calling with the latest number to a certain length, if 20 that would mean 20 individual queries?</p>
</li>
</ol>
<p>I ws hoping to use something like issueOrPullRequest(range of number)</p>
<p>Is that acceptable for graphQL? I was under the impression less queries were the best thing…</p>

**Comment by lee-dohm** [comment, marked as answer]
<p>You are correct, that you would have to make multiple queries. I agree that the multiple queries would probably be wasteful but it was the only way I could see of obtaining information similar to what is available in the REST API.</p>

**Comment by kttkjl** [comment]
<p>Mhm… I tried quite a bit to see what I can do to use the least amount of queries - but to no avail.</p>
<p>I’m guessing Github just decided to separate issues from pull requests for good in v4 to minimize extra data from coming in?</p>
<p>And thanks for your answers and patience!</p>
